### PR TITLE
getOutOfBagPredictions

### PR DESCRIPTION
**Pull request by PhilippPro**
This is a new PR for #902, created a branch, so all can work on it. 

I just transformed it, so that a standardized "predict" object is provided, so that e.g. performance can be applied on the resulting object.

Further development could be changing it, so it is usable for Tuning and Bagging Wrapper.

Code example:

```

 set.seed(123)
 training.set = sample(1:150, 50)
 lrn = makeLearner("classif.ranger", predict.type = "prob", predict.threshold = 0.6)
 mod = train(lrn, sonar.task, subset = training.set)
 oob = getOutOfBag(mod, sonar.task)
 oob
 performance(oob, measures = list(auc, mmce))

```

Renamed some of the functions. 

Still not completely done, e.g.:

> This generates NAs if there are NAs in t.col.

### CODE REVIEW

**Comment by PhilippPro** [comment]
changed the test for equality of task and model

File: R/RLearner_classif_randomForest.R [hunk 1]
```diff
@@ -50,3 +50,13 @@ predictLearner.classif.randomForest = function(.learner, .model, .newdata, ...)
   type = ifelse(.learner$predict.type=="response", "response", "prob")
   predict(.model$learner.model, newdata = .newdata, type = type, ...)
 }
+
+#' @export
+getOutOfBagPredictions.classif.randomForestSRC = function(.learner, .model) {
+  preds = .model$learner.model$predicted.oob
+  if (.learner$predict.type == "response") {
+    factor(colnames(preds)[max.col(preds)], levels = colnames(preds))
```
Comment (jakob-r) [inline R/RLearner_classif_randomForest.R]: comment about what is done here would be nice and helpful to understand

**Comment by berndbischl** [comment]
sorry,

1) pls rename: getOOBPredsLearner.classif.foo

2) pls rename: getOOBPreds for the main function

3) pls use roxygen tags for wrapped model, the task and so on

4)
st = system.time(p <-  getOutOfBagPredictionsModel(object$learner, object))
shouldnt this be NA?
also document this.

5) 
+  assertSetEqual(identical(task$task.desc, object$task.desc), TRUE) 
this we want in a helper function

6) pls put the test in helper_learners_all

**Comment by PhilippPro** [comment]
ready for review @berndbischl

**Comment by berndbischl** [comment]
1)

>> st = system.time(p <- getOutOfBagPredictionsModel(object$learner, object))
>> also document this.
you should document this also in ROXYGEN

2) 
p = getOOBPredsLearner(object$learner, object)
you need to check the return type here and its dimensions at least

3) 
have you documented what happens with probabilities? and maybe tested this?
its ok if we dont support this at the moment but it must be well defined

**Comment by PhilippPro** [comment]
to 2) I test the object like in predict with checkPredictLearnerOutput(object$learner, object, p)

**Comment by berndbischl** [comment]
@all i would like to draw this in quickly. otherwise it rots.

@PhilippPro 
can you please resolve the conflict. and what is the status now here?

@schiffner @larskotthoff @mllg @jakob-r @zmjones 
please help me review this.

**Comment by larskotthoff** [comment]
Looks good to me.

**Comment by zmjones** [comment]
why is R/checkModelCorrespondsTask.R in its own file? otherwise looks fine to me.

**Comment by PhilippPro** [comment]
conflicts resolved (hope for travis). We wanted to make this feasible with tuning, but I think I need your advise here and we can do it in another new branch then. 

Where should we put the check otherwise @zmjones? One could also put all checks in one file...

**Comment by larskotthoff** [comment]
@PhilippPro Looks like Travis is failing with some documentation warnings. Could you have a look at those please?

**Comment by zmjones** [comment]
@PhilippPro I'd put it in utils.R maybe?

**Comment by berndbischl** [comment]
> why is R/checkModelCorrespondsTask.R in its own file? otherwise looks fine to me.

this seems to be good to me. we have a couple of very similar "check" functions, all in their own file. 
so this is kinda in spirit with the code base

**Comment by zmjones** [comment]
@berndbischl maybe it would make sense to combine them when they are that small? it is the case that they are a bit underused right? collecting them might make them easier to find.

**Comment by larskotthoff** [comment]
I think they're actually easier to find in separate files -- name of function -> name of file.

**Comment by berndbischl** [comment]
> @berndbischl maybe it would make sense to combine them when they are that small? it is the case that they are a bit underused right? collecting them might make them easier to find.

i do see the point of grouping them in a well named and documented file. but please discuss this in another clean issue for that purpose.

**Comment by berndbischl** [comment]
@PhilippPro 
can you please fix the tracis issues? it doc stuff, thats missing

**Comment by PhilippPro** [comment]
I do not understand, why we have the distinction between object and model (e.g. in predict). Furthermore I discovered an empty R file (isFailureModel.R) that probably can be deleted? The corresponding functions are in WrappedModel now. This is one example of what I mentioned on the conference, that it can be difficult to find the corresponding functions when you do not know where they are... (and this was also a bit easier than other examples)

**Comment by PhilippPro** [comment]
problems in the pdf version of the manual says jenkins, what is the reason? Travis runs were ok...

**Comment by berndbischl** [comment]
> I do not understand, why we have the distinction between object and model (e.g. in predict).

but exactly do you mean here?

> Furthermore I discovered an empty R file (isFailureModel.R) that probably can be deleted? The corresponding functions are in WrappedModel.R now.

ok delete in then

> This is one example of what I mentioned on the mlr-workshop, that it can be difficult to find the corresponding functions when you do not know where they are... (and this was also a bit easier than other examples)

if you dont know how to find functions please dicuss it somewhere else. like in slack. or open an issue. and i can help you

**Comment by PhilippPro** [comment]
in predict.WrappedModel we take the "object" and assign "model = object". Why not just stay with "object"?

**Comment by berndbischl** [comment]
> in predict.WrappedModel we take the "object" and assign "model = object". Why not just stay with "object"?

somebody apparently just renamed the object to make it more readable. you can change and clean this up but not in this PR pls. i also not find it very relevant.

**Comment by berndbischl** [comment]
@PhilippPro 

1) please try to review your stuff a bit yourself, so we can speed up the process

2)  why are you changing the rox template definition for arg_wrappedmod?
that makes zero sense! please revert this!
and even your indendation is wrong, it should be 2 spaces in the 

3) checkModelCorrespondsTask.R
please add at least one comment line to describe what the function does. and the error msg is not really precise and helpful. say that the descs differ....

4) you still have wrong indentation (vertical arg aligment) in code. please change that and really reconfigure your rstudio.

5) before getOOBPredLearner, the generic please FULLY document its API. that means what goes in and what comes out! thats the most important part

**Comment by PhilippPro** [comment]
> somebody apparently just renamed the object to make it more readable. you can change and clean this up but not in this PR pls. i also not find it very relevant

It is relevant, because otherwise you have to create a new template for all the files where you use "object" instead of "model". (I did not do this now, because I think it would be better to just rename this and have one template for all)

**Comment by berndbischl** [comment]
please say that in getOOBPredsLearner the output should have extactly the same structure as in predictLearner. that should be enough

**Comment by PhilippPro** [comment]
merge @berndbischl ? It would be good so I could continue with the tutorial.

**Comment by jakob-r** [comment]
I will have a look on it.

**Comment by jakob-r** [comment]
Ah. My mistake. However the other checkfunctions are used more than once with the exception of `checkAggrBeforeResample.R` which is just used from `resample.R`.
So I would vote to keep it together in one file but I will accept a veto :)

**Comment by jakob-r** [comment]
Everything works now and looks fine to me.
Please check and merge @berndbischl @mllg @larskotthoff

**Comment by larskotthoff** [comment]
Looks good to me.

File: R/RLearner_classif_randomForestSRC.R [hunk 2]
```diff
@@ -65,6 +65,15 @@ predictLearner.classif.randomForestSRC = function(.learner, .model, .newdata, ..
...
 #' @export
+getOOBPredsLearner.classif.randomForestSRC = function(.learner, .model) {
+  preds = .model$learner.model$predicted.oob
+  if (.learner$predict.type == "response") {
+    factor(colnames(preds)[max.col(preds)], levels = colnames(preds))
+  } else {
+    preds
+  }
+}
+
```
Comment (schiffner) [inline R/RLearner_classif_randomForestSRC.R]: `@export`

File: R/RLearner_classif_ranger.R [hunk 3]
```diff
@@ -45,6 +45,10 @@ predictLearner.classif.ranger = function(.learner, .model, .newdata, ...) {
 }
 
 #' @export
+getOOBPredsLearner.classif.ranger = function(.learner, .model) {
+  .model$learner.model$predictions
+}
+
```
Comment (schiffner) [inline R/RLearner_classif_ranger.R]: `@export`

File: R/RLearner_regr_randomForestSRC.R [hunk 4]
```diff
@@ -62,6 +62,10 @@ predictLearner.regr.randomForestSRC = function(.learner, .model, .newdata, ...)
 }
 
 #' @export
+getOOBPredsLearner.regr.randomForestSRC = function(.learner, .model) {
+  as.numeric(.model$learner.model$predicted.oob)
+}
+
```
Comment (schiffner) [inline R/RLearner_regr_randomForestSRC.R]: `@export`?

File: R/RLearner_regr_ranger.R [hunk 5]
```diff
@@ -43,6 +43,10 @@ predictLearner.regr.ranger = function(.learner, .model, .newdata, ...) {
 }
 
 #' @export
+getOOBPredsLearner.regr.ranger = function(.learner, .model) {
+  .model$learner.model$predictions
+}
+
```
Comment (schiffner) [inline R/RLearner_regr_ranger.R]: `@export` ?

Comment (PhilippPro) [inline R/RLearner_regr_ranger.R]: They are not exported, but only used internally, I think?

File: R/getOOBPreds.R [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#'
+#' @description
+#' Learners like \code{randomForest} produce out-of-bag predictions. 
+#' \code{getOOBPreds} extracts this information from trained models and builds a 
+#' prediction object as provided by predict (with prediction time set to NA). 
+#' In the classification case: 
+#' What is stored exactly in the [\code{\link{Prediction}}] object depends
+#' on the \code{predict.type} setting of the \code{\link{Learner}}.
+#' 
+#' You can call \code{listLearners(obj = NA, properties = "oobpreds")} to get a list of learners
```
Comment (schiffner) [inline R/getOOBPreds.R]: `listLearners(properties = "oobpreds")` should suffice

File: R/RLearner_regr_ranger.R (hunk 5 above)
Comment (schiffner) [inline R/RLearner_regr_ranger.R]: It's needed to register the S3 methods in the NAMESPACE

**Comment by schiffner** [comment]
Looks good to me.

File: R/RLearner_classif_randomForest.R [hunk 7]
```diff
@@ -52,6 +52,14 @@ predictLearner.classif.randomForest = function(.learner, .model, .newdata, ...)
...
 
 #' @export
+getOOBPredsLearner.classif.randomForest = function(.learner, .model) {
+  if (.learner$predict.type == "response") {
+    unname(.model$learner.model$predicted)
+  } else {
+   .model$learner.model$votes
+  }
+}
+
```
Comment (jakob-r) [inline R/RLearner_classif_randomForest.R]: It's not visible in the diff but it's there :thinking:

Comment (jakob-r) [inline R/RLearner_classif_randomForest.R]: because it was *stolen* from the following method :laughing:

Comment (schiffner) [inline R/RLearner_classif_randomForest.R]: Really? Ups, I thought it was missing before `getFeatureImportanceLearner.classif.randomForest`

**Comment by jakob-r** [comment]
@schiffner Thanks for spotting the export mistake!

**Comment by schiffner** [comment]
Not a problem. Thank you and Philipp.
I think one export is still missing in RLearner_regr_randomForestSRC.R?

**Comment by jakob-r** [comment]
We will merge this into master after merging this together with [improved se estimation for random forests](https://github.com/mlr-org/mlr/pull/1420)

**Pull request merged**